### ISSUE TITLE
Where is the Dataloader for OBB data

### ISSUE THREAD

**Issue by dk-teknologisk-mlnn** [ultralytics/ultralytics#9287]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

In the original repo for yolov5 v 5-6 it was easy to follow the code from dataloader, loadimageasandlabels and see what happens when it reads the labels. 
But in the new code structure it is really messy and obscure. 
I cannot follow it from trying to cache the bounding boxes and how it uses " self.use_obb" to distinguish between xywh format and xy xy xy xy poly format and where theta comes into the equation.

I have used the yolov5obb repo and that follows the old structure, where it reads the poly and has a poly2xywht (theta) function. Its downfall is that it uses cv2 smallest rect function to find the enclosing rectangle . and if W~= H it is randomly assigned +/- 90 degrees. I want to see if we have that problem here?

I'd also like to see what it takes to have it read txt files that directly put cx, cy, w,h, theta so we dont risk having it compute a different theta than we used in annotation. and also look into 360 degree thetas.

short version: in which file and line number does it read obb poly and convert to xywht?

**Comment by glenn-jocher** [comment]
@dk-teknologisk-mlnn hello! Thanks for reaching out with your question. 🌟 

The handling of Oriented Bounding Boxes (OBB) in YOLOv8 can indeed look a bit complex at first glance because of the evolved code structure aimed at providing more flexibility and performance.

For your specific question on how OBB data is loaded and processed, the OBB handling is integrated within the dataset loading mechanism:

- In YOLOv8, the parsing and conversion of OBB data, including reading polygonal (xy xy xy xy) formats and conversion to the center coordinates + width, height, and angle format (cx, cy, w, h, theta), occur within the dataset class.

- The functionality you're looking for, particularly where the dataset decides on using `self.use_obb` to distinguish between different bounding box formats and apply conversions, is implemented in the dataset's `__getitem__` method, which is typically found in the file handling the dataset loading. 

- Unfortunately, without direct access to inspect the most recent repository structure, I can't provide an exact file name or line number. However, it would be within a file related to data loading operations, possibly named similarly to `datasets.py` or within a module dedicated to data handling.

Regarding direct handling of annotations in the format cx, cy, w, h, theta -- if this specific functionality isn't currently supported directly by the codebase, it might necessitate a customization of the dataset loading methods. Custom parsing functions can be added to interpret your specific annotation format and conduct necessary transformations.

For 360-degree theta handling, careful additional handling could be required to ensure consistent interpretation and conversion during both training and inference phases.

For contributions or suggestions on adding or improving these functionalities, engaging in the repository's discussions or opening a feature request issue would be the best approach to bring this to the team's attention.

I hope this provides a starting point for navigating OBB handling in YOLOv8. If you have further specific questions or need assistance, the community and contributors on the GitHub issues and discussions pages are great resources. Happy coding! 😊

**Comment by dk-teknologisk-mlnn** [comment]
The thing is in your old repo (yolov5) __getitam_ has  all the code to and I can trace everything and I've adapted it to read read my data as it comes out of my annotator and added my on augmentations as well. 
In this current one getitem is pretty much empty. and it just returns a function call that I follow and it gets really obscure and it passes around use_obb, and magically boxes appear [with no code to actually open and read the txt files to be found] and gets passed around that the comment state are xywh format. 

I've looked and searched through dataset/py and its parent class base.py and augmentation.py. I happily found where the forced albumentation gets added so I out-commented that one [this should really be controllable in the hyp.yaml like all the other augments]. 

I searched every call to "cv2" and only imread and resize are used in dataset classes. and I searched for "open(" and only txt files that has image lists are opened in dataset classes. 

I guess I have to wrap it around a visual studio project and use debug mode to F10 through each and every line in runtime.

**Comment by dk-teknologisk-mlnn** [comment]
ok found the conversion in ops.py and made a div 0 error to have it trace it:

It uses cv2.minAreaRect(pts) . btw the function comment is wrong saying rotation has to be degrees 0 to 90. 
Maybe it is enough to define the angle from the vector made from P0 -> P1 to handle square boxes and maybe also to get 360 degree angles in this conversion function. the box -> poly should handle those two other quadrants correctly. Have to check the loss & IOU computation.

Original Traceback (most recent call last):
  File "C:\Anaconda3\envs\pytorch\lib\site-packages\torch\utils\data\_utils\worker.py", line 287, in _worker_loop
    data = fetcher.fetch(index)
  File "C:\Anaconda3\envs\pytorch\lib\site-packages\torch\utils\data\_utils\fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "C:\Anaconda3\envs\pytorch\lib\site-packages\torch\utils\data\_utils\fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "C:\Python\dtideepnet\DTIInstanceSegNet\ultralytics\data\base.py", line 253, in __getitem__
    return self.transforms(self.get_image_and_label(index))
  File "C:\Python\dtideepnet\DTIInstanceSegNet\ultralytics\data\augment.py", line 74, in __call__
    data = t(data)
  File "C:\Python\dtideepnet\DTIInstanceSegNet\ultralytics\data\augment.py", line 943, in __call__
    xyxyxyxy2xywhr(torch.from_numpy(instances.segments)) if len(instances.segments) else torch.zeros((0, 5))
  File "C:\Python\dtideepnet\DTIInstanceSegNet\ultralytics\utils\ops.py", line 532, in xyxyxyxy2xywhr
    is_torch = isinstance(corners, torch.Tensor) /0
ZeroDivisionError: division by zero

**Comment by dk-teknologisk-mlnn** [comment]
Nope, because for some obscure reason PTS is not the 4 xy points given in the label file, but magically it got turned into 100 points!

**Comment by glenn-jocher** [comment]
@dk-teknologisk-mlnn hey there! Thanks for diving deep into the details and sharing your findings. 🕵️ 

Regarding the resampling of points to 100 for Oriented Bounding Boxes (OBB), it was intended to uniform the representation of polygons across the board for consistency in processing. However, you bring up a good point about potentially affecting training with changes to the `segment_resamples`. 

In the context of your annotator's format and concerns about the angle handling in the loss, you're right that the IOU and subsequent losses might not heavily penalize small deviations in angle, especially for squares or similar conditions you mentioned. These are insightful observations and definitely worth further inspection!

For adjusting how angles are factored in, or if you're considering a custom approach to better represent your unique dataset requirements, it seems you're already on a promising path by experimenting with resample points and considering angle weights in loss calculations. Careful adjustments and testing will be key to fine-tuning performance here.

If resampling to the exact points from your annotations works without negatively impacting other areas of training, it could be a viable solution. Just keep an eye on overall model performance to ensure no unintended consequences.

We're always looking to improve, and feedback like this is invaluable. Keep those questions and suggestions coming! 👍

**Comment by dk-teknologisk-mlnn** [comment]
I just trained with resample = 5 , and disabled patience to allow training for 2000 epochs more after the metrics stop improving, but the box loss keeps improving. Also realized I should not use flip in augment! I also made the images square so that rect corners that rotate out of the image can be there in the black part. now the predictions are actually pretty good.

**Comment by glenn-jocher** [comment]
@dk-teknologisk-mlnn-mlnn That's fantastic to hear! 🎉 It sounds like the adjustments you've made, particularly with the resample = 5 and modifications to the augmentation strategy, are paying off. It's great that you're seeing continuous improvement in the box loss and getting pretty good predictions now.

Disabling the patience and allowing for extended training epochs is an interesting strategy, especially for complex datasets or when fine-tuning angles and orientation in detection tasks. Making the images square to accommodate rotating corners is a clever approach to maintain context within the frame bounds.

Thanks for sharing your progress and insights. This kind of experimentation and adaptation is crucial to overcoming specific challenges in object detection projects. Keep up the fantastic work! If you come across any more interesting findings or have further questions, feel free to ask. Happy detecting! 🚀
